maintainers/scripts/update.nix: Add support for auto-commiting changes

jtojnar opened the pull request.

jtojnar committed:
update.nix: use ThreadPoolExecutor

Not sure why I chose ProcessPoolExecutor in the first place.

jtojnar committed:
common-updater-scripts: add --print-changes flag

Printing the changed file and new version can be used to commit the changes to git.

jtojnar committed:
maintainers/scripts/update.nix: Add support for auto-commiting changes

Update scripts can now declare features using

	passthru.updateScript = {
	  command = [ ../../update.sh pname ];
	  supportedFeatures = [ "commit" ];
	};

A `commit` feature means that when the update script finishes successfully,
it will print a JSON list like the following:

	[
	  {
	    "attrName": "volume_key",
	    "oldVersion": "0.3.11",
	    "newVersion": "0.3.12",
	    "files": [
	      "/path/to/nixpkgs/pkgs/development/libraries/volume-key/default.nix"
	    ]
	  }
	]

and data from that will be used when update.nix is run with --argstr commit true
to create commits.

We will create a new git worktree for each thread in the pool and run the update
script there. Then we will commit the change and cherry pick it in the main repo,
releasing the worktree for a next change.

jtojnar committed:
gnome3.updateScript: implement commit feature

jtojnar commented:
The main issue with this occurs when multiple `updateScript`s change the same file.

I would like to create a pool of the `worktrees` and need something that would pair `ThreadPool` with the `WorkTreePool` so that I could handle the thread in `as_completed` but not release the used worktree until a commit is cherry-picked.

Alternately, I could create a worktree per update but that would be unnecessarily wasteful. Though, this method would not require the `updateScript`s to print the modified files – we could just `git commit -A`.

worldofpeace commented on doc/stdenv.xml:
```diff
@@ -709,14 +709,50 @@ passthru.updateScript = writeScript "update-zoom-us" ''
+       </term>
+       <listitem>
+        <para>
+         The update script will print a JSON list containing an containing an object for each updated attribute whenever it exits with <literal>0</literal> return status. Empty list can be returned when the script did not update any files: for example, when the attrite is already latest version. The required keys can be seen below:
```
```suggestion
         The update script will print a JSON list containing an object for each updated attribute whenever it exits with <literal>0</literal> return status. Empty list can be returned when the script did not update any files: for example, when the attribute is already the latest version. The required keys can be seen below:
```

NeQuissimus commented:
Could you then go through all the `update.sh` files and remove the git commands? :)
I like this, I've been trying to streamline the update process for packages I am interested in.
We should probably also have the ability to attach derivations to be built and tests to be run as part of the updating process. And revert any changes if the builds/tests fail (or maybe it can already do that and I am using it wrong :D).

jtojnar commented:
> Could you then go through all the `update.sh` files and remove the git commands? :)

I think I removed most of them when we parallelized the updater.

> We should probably also have the ability to attach derivations to be built and tests to be run as part of the updating process. And revert any changes if the builds/tests fail (or maybe it can already do that and I am using it wrong :D).

In GNOME, we use this for the menial parts of update process, expecting that the package might not build. The failures (most often dependency changes) need to be fixed manually but we still want to keep the commit.

jtojnar commented:
I have managed to get it to pretty much the state I wanted: It will create a new git worktree for each thread in the pool and run the update script there. Then it will commit the change in the worktree and cherry pick it in the main repo, releasing the worktree for a next change.

jtojnar commented:
One thing to consider: do we want to updaters to have to add `commit` to supported features, or should we parse the old and new versions from the expression in `update.nix`?

FRidh commented on maintainers/scripts/update.py:
```diff
@@ -31,15 +40,35 @@ def main(max_workers, keep_going, packages):
+                        thread_name = package['thread']
+                        worktree, lock = temp_dirs[thread_name]
+                        changes = json.loads(p.stdout)
+                        for change in changes:
```
too much nesting. Consider a small separate function for the following lines

jtojnar replied:
I would say a much more pressing issue is how tightly coupled is everything with console input/output and the interruption not working very well.

Something like this looks better but interruption is still not handled gracefully. Perhaps using `queue.Queue` would make it nicer.

```patch
--- a/maintainers/scripts/update.py
+++ b/maintainers/scripts/update.py
@@ -9,6 +9,7 @@
 import threading
 
 updates = {}
+temp_dirs = {}
 
 thread_name_prefix='UpdateScriptThread'
 
@@ -28,6 +29,41 @@
 
     return subprocess.run(package['updateScript'], stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True, cwd=worktree)
 
+def update_packages(packages, max_workers, commit):
+    with contextlib.ExitStack() as stack, concurrent.futures.ThreadPoolExecutor(max_workers=max_workers, thread_name_prefix=thread_name_prefix) as executor:
+
+        if commit:
+            for i in range(max_workers):
+                temp_dirs[f'{thread_name_prefix}_{str(i)}'] = stack.enter_context(tempfile.TemporaryDirectory()), threading.Lock()
+
+            for wt, _lock in temp_dirs.values():
+                subprocess.run(['git', 'worktree', 'add', wt], check=True)
+
+        for package in packages:
+            updates[executor.submit(run_update_script, package, commit)] = package
+
+        for future in concurrent.futures.as_completed(updates):
+            package = updates[future]
+
+            try:
+                p = future.result()
+                if commit and 'commit' in package['supportedFeatures']:
+                    thread_name = package['thread']
+                    worktree, lock = temp_dirs[thread_name]
+                    changes = json.loads(p.stdout)
+                    for change in changes:
+                        subprocess.run(['git', 'add'] + change['files'], check=True, cwd=worktree)
+                        commit_message = '{attrName}: {oldVersion} → {newVersion}'.format(**change)
+                        subprocess.run(['git', 'commit', '-m', commit_message], check=True, cwd=worktree)
+                        subprocess.run(['git', 'cherry-pick', os.path.basename(worktree)], check=True)
+                yield package, True, None
+            except subprocess.CalledProcessError as e:
+                yield package, False, e.stdout
+            finally:
+                if commit and 'commit' in package['supportedFeatures']:
+                    lock.release()
+
+
 def main(max_workers, keep_going, commit, packages):
     with open(sys.argv[1]) as f:
         packages = json.load(f)
@@ -43,49 +79,22 @@
         eprint()
         eprint('Running update for:')
 
-        with contextlib.ExitStack() as stack, concurrent.futures.ThreadPoolExecutor(max_workers=max_workers, thread_name_prefix=thread_name_prefix) as executor:
-            global temp_dirs
-
-            if commit:
-                temp_dirs = {f'{thread_name_prefix}_{str(i)}': (stack.enter_context(tempfile.TemporaryDirectory()), threading.Lock()) for i in range(max_workers)}
-
-                for wt, _lock in temp_dirs.values():
-                    subprocess.run(['git', 'worktree', 'add', wt], check=True)
-
-            for package in packages:
-                updates[executor.submit(run_update_script, package, commit)] = package
-
-            for future in concurrent.futures.as_completed(updates):
-                package = updates[future]
-
-                try:
-                    p = future.result()
-                    if commit and 'commit' in package['supportedFeatures']:
-                        thread_name = package['thread']
-                        worktree, lock = temp_dirs[thread_name]
-                        changes = json.loads(p.stdout)
-                        for change in changes:
-                            subprocess.run(['git', 'add'] + change['files'], check=True, cwd=worktree)
-                            commit_message = '{attrName}: {oldVersion} → {newVersion}'.format(**change)
-                            subprocess.run(['git', 'commit', '-m', commit_message], check=True, cwd=worktree)
-                            subprocess.run(['git', 'cherry-pick', os.path.basename(worktree)], check=True)
-                    eprint(f" - {package['name']}: DONE.")
-                except subprocess.CalledProcessError as e:
-                    eprint(f" - {package['name']}: ERROR")
-                    eprint()
-                    eprint(f"--- SHOWING ERROR LOG FOR {package['name']} ----------------------")
-                    eprint()
-                    eprint(e.stdout.decode('utf-8'))
-                    with open(f"{package['pname']}.log", 'wb') as f:
-                        f.write(e.stdout)
-                    eprint()
-                    eprint(f"--- SHOWING ERROR LOG FOR {package['name']} ----------------------")
-
-                    if not keep_going:
-                        sys.exit(1)
-                finally:
-                    if commit and 'commit' in package['supportedFeatures']:
-                        lock.release()
+        for package, status, output in update_packages(packages, max_workers, commit):
+            if package['status']:
+                eprint(f" - {package['name']}: DONE.")
+            else:
+                eprint(f" - {package['name']}: ERROR")
+                eprint()
+                eprint(f"--- SHOWING ERROR LOG FOR {package['name']} ----------------------")
+                eprint()
+                eprint(output.decode('utf-8'))
+                with open(f"{package['pname']}.log", 'wb') as f:
+                    f.write(output)
+                eprint()
+                eprint(f"--- SHOWING ERROR LOG FOR {package['name']} ----------------------")
+
+                if not keep_going:
+                    sys.exit(1)
 
         eprint()
         eprint('Packages updated!')
```

jtojnar replied:
I moved it to separate function. I am also thinking about switching from `ThreadPoolExecutor` to asyncio and https://pypi.org/project/asyncio-pool/

jtojnar committed:
maintainers/scripts/update.nix: refactoring

Get rid of some globals, split main into smaller functions, rename some variables, add typehints.

hedning commented:
Tested this out and it seems to work, didn't seem to take a huge a amount of time compared to the previous version. It did get stuck on one update for a while, though I think that could happen before too.

One issue, I had to run `git worktree prune` after aborting an update once as the updater tried to reuse a non-existing worktree. Are we cleaning up properly on exit/start?

jtojnar commented:
> It did get stuck on one update for a while, though I think that could happen before too.

Do you mean that only single update script was running at the time? I guess we could have the threads waiting for their worktree lock to be released but that should not take a long time.

The context manager cleans worktrees once it finishes and should do the same even when exception occurs. But maybe exceptions triggered by signals are somewhat different, especially with regards to threads.

Maybe we need to handle the signals separately: https://stackoverflow.com/questions/18499497/how-to-process-sigterm-signal-gracefully

Or we could even use asyncio instead of threads. The main benefit is that unlike threads, the [subprocesses](https://docs.python.org/3/library/asyncio-subprocess.html#asyncio.asyncio.subprocess.Process) can be stopped easily. Perhaps we could even have a [pool](https://pypi.org/project/asyncio-pool/) of worktrees, simplifying the code even more. https://stackoverflow.com/questions/41677434/how-to-make-a-asyncio-pool-cancelable

This series offers a nice description of how python does these things: https://pymotw.com/3/concurrency.html

jtojnar commented:
Some more questions:

We probably need to use the worktrees for non-commit use case as well, right? Otherwise, the `packages.nix` (for example Vala) could be broken, since `update-source-version` will race.

How would we merge the changes then, though? `git add -A` → `git commit` (versions would be obtained from derivation name before and after the change) → `git cherry-pick --no-commit`? What would happen when there are conflicts git cannot resolve? (I really need to focus on possible failure modes in the current solution as well.)

Or should we still update the files in-tree and try to solve this by locking? How would we know what files to lock? Another attribute listing the locked files in the `updateScript` attrset?

And since I am already considering always using the worktree, we could get the commit feature for free (the merge procedure would only differ by the presence of `--no-commit` flag when cherry-picking). We would have to assume that each `updateScript` only updates a single attribute but we already do that – we only cherry-pick the last commit. Should we then get rid of the explicit commit feature? Do we remove the `--print-changes` from `update-source-version` as well?

Finally, should we reset the worktree after each merge? It would make conflicts always manifest, not only when they occur in different worker threads, which might be preferable.

jtojnar commented:
Also should the `commit` flag default to `true`?

hedning commented:
> Do you mean that only single update script was running at the time?

I think it all the other threads were done, think it was just one update that was stuck a while, probably unrelated.

> The context manager cleans worktrees once it finishes and should do the same even when exception occurs

Yeah, the directories in tmp were gone, but git still had references to them, it looked like the update script tried to reuse them and failed since they were no longer there. I might've pressed ctrl-c twice, killing the cleanup before it was entirely done though.

> Also should the commit flag default to true?

I wouldn't mind, it's just as easy (if not easier) to deal with the commits as un-committed changes in the worktree.

hedning commented:
When updating a single derivation the script still prepares `n` workstrees. It should at least limit the threads/worktrees to the number derivations being updated.

jtojnar commented:
Yeah, hopefully the Pool implementation will default to creating instances on demand.

hedning commented:
Hmm, it looks like it four instances on my machine even if there's only one package:
```
Going to be running update for following packages:
 - gnome-builder-3.34.0

Press Enter key to continue...

Running update for:
Preparing worktree (new branch 'update-tmphlopysqe')
Updating files: 100% (19962/19962), done.
HEAD is now at 5610389e1ff fixup! gnome3.gdm: 3.34.0 → 3.34.1
Preparing worktree (new branch 'update-tmpbk29udcw')
HEAD is now at 5610389e1ff fixup! gnome3.gdm: 3.34.0 → 3.34.1
Preparing worktree (new branch 'update-tmpy36__ucd')
HEAD is now at 5610389e1ff fixup! gnome3.gdm: 3.34.0 → 3.34.1
Preparing worktree (new branch 'update-tmpip9rlkpa')
HEAD is now at 5610389e1ff fixup! gnome3.gdm: 3.34.0 → 3.34.1
 - gnome-builder-3.34.0: UPDATING ...

```

jtojnar commented:
Yeah, I have not pushed the [new Pool implementation](https://github.com/NixOS/nixpkgs/pull/59372#issuecomment-485818022) yet. Hence the WIP status.

jtojnar committed:
python3Packages.asyncio-pool: init at 0.5.1

jtojnar commented:
Replaced by #98304
